Random removal, greedy insert

EmielK opened the pull request.

EmielK committed:
Random_removal, random_repair

EmielK committed:
random_removal, random_repair

EmielK committed:
Random removal

EmielK committed:
random_removal changes

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -51,3 +51,16 @@ def _plan_cost(self, problem: Problem) -> float:
+        """
+        self.customers.remove(customer)
+
+        for stacks in self.plan:
```
This is dog slow. It's much better to do the following:

1. Determine the current index of customer in self.customers, and name it `idx`. Then:
    - All stacks up to `idx` have `d<customer>`, the customer's delivery item.
    - All stacks after `idx` have `p<customer>`, the customer's pickup item.
2. For each stack up to `idx`, remove `Item(problem.demands[customer], DEPOT, customer)` using `find_stack` on `Stacks`.
3. For each stack after `idx`, remove `Item(problem.pickups[customer], customer, DEPOT)`, again using `find_stack`.

N-Wouda replied:
`find_stack` returns the stack containing the given item, in time `O(num_stacks)`. For this stack you will need to remove the item (there is no code for this yet, I believe, but a remove function on `Stack` should suffice).

N-Wouda commented on heuristic/destroy_operators/Random_removal.py:
```diff
@@ -0,0 +1,25 @@
+    for customer in random_state.choice(destroyed.problem.num_customers,
+                                        number_to_remove,
+                                        replace=False):
+
```
No need for white space.

N-Wouda commented on heuristic/functions/customers_to_remove.py:
```diff
@@ -0,0 +1,7 @@
+
+
+def customers_to_remove():
+    return int(Problem.num_customers * DEGREE_OF_DESTRUCTION)
```
You will need to pass a problem instance in - `Problem` is a class, not an instance. The class doesn't know about `num_customers`.

N-Wouda replied:
Use `problem`.

N-Wouda commented on heuristic/repair_operators/Random_repair.py:
```diff
@@ -0,0 +1,24 @@
+    unassigned list and places them in a random location in a random route.
+    Returns the resulting solution.
+    """
+    solution = old_solution.copy()
```
Repair operators do not need to copy.

N-Wouda commented on heuristic/repair_operators/Random_repair.py:
```diff
@@ -0,0 +1,24 @@
+    """
+    solution = old_solution.copy()
+
+    while solution.unassigned:
```
```python
while len(solution.unassigned) != 0:
```

EmielK committed:
random removal changes

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
+
+        for stacks in self.plan[:(idx - 1)]:
+            stack = stacks.find_stack(delivery_item)
+            stack.stack.remove(delivery_item)
```
This is not sufficient to remove the item - it also needs to be removed from the set member. Write a `remove` wrapper function on `Stack`.

N-Wouda commented on heuristic/destroy_operators/Random_removal.py:
```diff
@@ -0,0 +1,23 @@
+    and removes them from the solution.
+    """
+    destroyed = current.copy()
+    number_to_remove = customers_to_remove(Solution.problem)
```
This does not work

N-Wouda replied:
Use `current`.

N-Wouda replied:
Or `destroyed`, both would work.

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
+        or origin from the loading plan.
+        """
+        idx = self.customers.index(customer)
+        delivery_item = Item(Problem.demands[customer], DEPOT, customer)
```
This does not work.

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
+        """
+        idx = self.customers.index(customer)
+        delivery_item = Item(Problem.demands[customer], DEPOT, customer)
+        pickup_item = Item(Problem.demands[customer], customer, DEPOT)
```
Same here.

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -51,3 +52,22 @@ def _plan_cost(self, problem: Problem) -> float:
+        delivery_item = Item(Problem.demands[customer], DEPOT, customer)
+        pickup_item = Item(Problem.demands[customer], customer, DEPOT)
+
+        for stacks in self.plan[:(idx - 1)]:
```
Be _very_ careful about the offsets here - I do not know them by heart, and you'll need to test this to find out if this is the correct cut-off point. I think it's too early, but again, I could be wrong.

EmielK replied:
In the graphs, at idx there was only a pickup present, which is why I thought this would be the cut-off point.

N-Wouda replied:
Alright then!

N-Wouda replied:
The first loading plan is always just after leaving the depot, so this should run up to `idx + 1`, not `idx - 1`). Similarly for the pickup item, which should be in the tour from `idx + 1` onwards.

N-Wouda replied:
OK I need to think about this a little more.

N-Wouda replied:
Yeah it's up to `idx + 1` for deliveries, and from `idx + 1` onwards for pickups. See e.g. [this build](https://travis-ci.com/N-Wouda/OR-Analysis/jobs/288094552) where I have updated the validator to consider this.

N-Wouda committed:
Merge branch 'master' into random-greedy

EmielK committed:
random removal changes

N-Wouda committed:
Auto stash before merge of "random-greedy" and "origin/random-greedy"

N-Wouda committed:
Fix route customer removal

N-Wouda committed:
Docstring

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,45 @@
+    #  per unassigned probeer in te voegen per route op elke plek en bereken
+    #  kosten, en ook kosten voor nieuwe route. Selecteer laagste en plaats daar
+
+    while current.unassigned != 0:
```
```python
len
```

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,45 @@
+    #  kosten, en ook kosten voor nieuwe route. Selecteer laagste en plaats daar
+
+    while current.unassigned != 0:
+        customer = next(customer for customer in current.unassigned)
```
```python
customer = current.unassigned.pop()
```

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,45 @@
+
+        # Closest visitable node.
+        nearest = np.argmin(current.problem.distances[assigned, customer])
+        # Find route it is in and insert customer after closest other customer.
```
So I guess the plan here is the following:

```
IF nearest is the depot THEN
  Make route for customer
ELSE IF customer can be assigned to route containing the nearest customer THEN
  Assign customer to this route and update the loading plans
ELSE
  Make route for customer
```

EmielK committed:
random greedy, (WIP)

N-Wouda committed:
Merge branch 'master' into random-greedy

N-Wouda committed:
Some fixes

N-Wouda committed:
Fix solution reading on validator

N-Wouda committed:
Some clean-up for greedy_insert

N-Wouda commented on heuristic/classes/Stacks.py:
```diff
@@ -23,6 +25,10 @@ def __iter__(self):
         return self.stacks[idx]
 
+    def copy(self):
+        tmp = copy.deepcopy(self)
```
Or just ```return copy.deepcopy(self)```. Also types.

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    TODO.
+    """
+    while len(current.unassigned) != 0:
```
This entire method is _far too slow_. Can you estimate the run-time cost?

N-Wouda committed:
Simplify remove_customer

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,76 @@
+
+            # insert for delivery in all stacks before customer
+            for plan in route.plan[:idx_customer + 1]:
+                plan.shortest_stack().push_rear(delivery)
```
There is no guarantee the tour capacities are respected when you do this.

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,76 @@
+
+            # insert pickup for all stacks after customer
+            for plan in route.plan[idx_customer + 1:]:
+                plan.shortest_stack().push_rear(pickup)
```
Same here.

N-Wouda committed:
Style fixes

N-Wouda committed:
Use numpy slicing to determine routing costs

N-Wouda committed:
Fix distance computation order

This does not matter much for symmetric distance matrices, but would for asymmetric.

N-Wouda committed:
Random route removal

N-Wouda committed:
Fix Stacks cost computation

EmielK committed:
feasibility check

N-Wouda committed:
Fix volume computation on Stack objects

N-Wouda committed:
Fix mistaken cost estimates

N-Wouda committed:
Better big-O remove_customer cost description

N-Wouda committed:
Constant for number of iterations

EmielK committed:
Auto stash before merge of "random-greedy" and "origin/random-greedy"

EmielK committed:
Removed M from constants

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
 
         return cost
+
+    def is_feasible(self, problem: Problem, customer: int) -> bool:
```
I'd probably call this ```can_insert```

N-Wouda replied:
Also make ```problem``` the last argument, not the first.

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
+        """
+        Checks if inserting a customer in this route is feasible.
+        """
+        delivery = Item(problem.demands[customer], DEPOT, customer)
```
```d_volume```, a `float` of just the delivery volume. No need for an Item here. Same for pickup, `p_volume`

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
+        delivery = Item(problem.demands[customer], DEPOT, customer)
+        pickup = Item(problem.pickups[customer], customer, DEPOT)
+
+        if all(stacks.shortest_stack().volume() + delivery.volume <=
```
Difficult to read. Split it into can_deliver and can_pickup, two boolean variables, and return `can_deliver and can_pickup`.

N-Wouda commented on heuristic/classes/Route.py:
```diff
@@ -56,3 +56,72 @@ def _plan_cost(self, problem: Problem) -> float:
+
+        return result
+
+    def opt_insert(self, problem: Problem, customer: int) -> int:
```
Can this not re-use `insert_cost()`?

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,64 @@
+        # Costs new route
+        cost_new_route = ([current.problem.distances[DEPOT + 1, customer + 1] +
+                           current.problem.distances[customer + 1, DEPOT + 1]])
+        heapq.heappush(insertion_heap, (cost_new_route, 0, len(current.routes)))
```
Why does this need to be on the heap? Can't you compare this to the routes on the heap to determine if it's worthwhile? That'd allow you to store routes on the heap, which is just that much easier.

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    TODO.
+    """
+
```
Remove whitespace

N-Wouda commented on heuristic/repair_operators/greedy_insert.py:
```diff
@@ -0,0 +1,64 @@
+
+def greedy_insert(current: Solution, rnd_state: RandomState) -> Solution:
+    """
+    TODO.
```
And add a description. This is related to `sequential best insertion` in Hornstra et al. (2020).

N-Wouda replied:
Including a run-time complexity!

N-Wouda committed:
Nearest pairs removal

N-Wouda committed:
Clean-up some stuff in greedy_insert

N-Wouda committed:
Some further changes to structure

N-Wouda committed:
Some misc. fixes to program structure

N-Wouda commented:
Mostly OK, LGTM.
